Feature/inventory system

cayb0rg opened the pull request:
# Adventure Inventory System

Creators can create their own items and select or upload icons to these items.

Creators can select items to give players upon visiting a node.

Creators can require certain items for answer choices to be selectable. 

The widget will let the player know they are missing an item. 

Players receive an alert when they receive an item and can view their inventory by clicking on the backpack icon.

### Other changes in this PR include:
- Node editor windows are made more responsive and roomier
- Player converted to flex containers
- Short answer string matching now has the option to be case sensitive and/or special character sensitive
- Updates to MWDK3

### What's Left
- ~~Player Accessibility~~
  - ~~Hide background and other options when modals are open~~
  - ~~Add aria-labels to buttons and other inventory items~~
  - ~~Add aria-live updates for new and removed inventory items~~
- ~~Ensure scoring works with MWDK3~~
- Update creator's and player's guides to match latest style changes

clpetersonucf commented:
It'll take me a little bit to run through everything, but I wanted to start working on feedback items asap. A few suggestions so far:
- The required items UI (for multiple choice and hotspot) and the "give player items" options in the creator can probably be hidden when the inventory is empty, as a way to visually "opt out" of the feature if the user chooses not to interact with it.
- Similarly, the "Gives item(s)" mouseover tooltip when hovering over a node shouldn't be visible unless it actually gives an item.
- I like the Add Item UI in the creator, but it's not immediately obvious that I can further edit the item once I create it. Could the edit item dialog items become visible by default once you enter a name?
- The Required Items UI will need a polish pass or two, but I'm not sure how to improve it without mocking up a few ideas. I'll follow up on this one.
- Short Answer nodes in the creator don't appear to have the Required Items UI, but I think I agree that the combination of its existing requirements *plus* required items may be too complicated. This isn't really a feedback item, just acknowledging the decision to not include Require Items for Short Answer nodes.
- Icons are not provided with new items by default, but the player will crash if an icon is not present.
- The answer choice UI for multiple choice nodes in the creator does not appear to scroll far enough and causes occlusion issues with > 1 answer choices.
- The flex enhancements to node edit screens in the creator are terrific. It's clear we're still dealing with some significant vertical content compression issues though. I wonder if we can add another responsive option, for wide enough window width, that splits the screen in half with question/asset content management on the left and answer content management on the right (for MC and SA node types)?

cayb0rg commented:
Thank you so much for the feedback! For the Required Items UI, perhaps a simple "Require Items" button next to Choose Destination that opens a modal would suffice (and would appear only if there are items). This could also make it simpler to include with the Short Answers.

clpetersonucf reviewed (changes requested):
This is continuing to look more refined. Some high-level design feedback:

- The "Add Items"/"Edit Items" on the base UI could be made a little more obvious and stateful. I'd like it to maybe indicate how many items are created for the inventory system, like "Edit Items (5)", and include the backpack icon used elsewhere.

- The Required Items UI for MC is looking good so far. I appreciate how it opens on the right side for wide resolutions, and atop the answer section when in narrower resolutions. I wouldn't mind the header specifying which answer the Required Items UI is currently associated with: probably going by the answer text.

- The actual Required Items icon to open the window still needs work... I'm just not sure how yet. I think the lock/unlock icon is good, but we can refine it further. I'll have to see about sketching up some ideas.

- The "Add Items"/"Edit Items" button could probably be changed slightly to more clearly specify how many items, like the one on the base UI.

- I'd love small icons to be positioned on the nodes and arrows to indicate whether an inventory system state change is present. I could see three types: when an item is being given to the user when they visit a node, when an item is removed from the user when they visit a node, and whether a specific answer choice requires an item in order to be selected (probably positioned on the arrow from the parent to the child node)

- I think there's promise in the changes made to the hotspot UI, but it's not quite there yet. I'm not wild about the toolbar options sort of floating in space on the left side for wide resolutions. We can tighten that up a bit and move the image + hotspot canvas to the left a little more. I really like the narrower variation, and how it floats on top of the image canvas at the bottom. Making the node management dialogs draggable is a big improvement, and I appreciate how the required items dialog is attached and moves with it. When the required items dialog displays on the right, for wider resolutions, it probably needs to be more clearly associated with the hotspot being edited, much like my previous comment for the MC node editing screen. Again, this is just visually tightening stuff up a bit. The bones are all there and it looks good.

Other, not design-related comments:
- Are the icomoon fonts required, now that we're using Simon's icons instead of font glyphs? If not, we could clean those out as well as the `liga.js` library and associated stylesheet.
- The `load_icons.sh` script is really neat, and I'm impressed that you're using it. I'm torn on keeping it in the git history - it can probably be there, but I might include some comments at the top of the script to specify exactly what it's for.

clpetersonucf reviewed (changes requested):
Some additional feedback from testing:

- It might be useful for the inventory icon in the player to display the total count of items inside it. Additionally, maybe a hover state on mouseover to indicate what you have available.
- The actual inventory screen (in the player) can probably be enlarged a bit. I'd also like to see a quantity next to each item, for those with an amount > 1. It's a little annoying having to click on each one to see how many I have.
- The "add items"/"edit items" UI in the creator no longer specifies that you can remove an item from an inventory. Setting the quantity to -1 works, but we could just as easily split this UI in two and have a "remove items" section below the "give items" section to make it clearer that it's an option. Similarly, the "gives items" tooltip shows a negative quantity when an item is removed - while technically true, it should probably have a version that says "removes items: X" in addition to the "gives items: X" tooltip.
- Removed items remain in the inventory (in the player) with a quantity of 0.
- I like the tooltip in the player when an item is added or removed - I think we can enhance it a bit to make it more obvious, kinda like the toasts in the creator, and perhaps provide the item's icon as well. It should be very obvious when the user is given an item or an item is removed.
- The question text textarea (in the creator) no longer changes size depending on the node type being edited - for MC, hotspot, and short answer the default size makes sense, but the text area can safely take up a lot more real estate for narrative and end points. Additionally, if I change its size, that resize persists to other node types, potentially pushing things out of frame.
- Thinking about the hotspot editor in the creator some more: what if we moved the question text to the left of the image/hotspot canvas and gave that canvas the full height of the editor when in wider resolutions?

cayb0rg commented:
Thank you for the feedback!

Changes:
- Added circles around item icons surrounding individual nodes to indicate whether they are being taken or given
- Updated hotspot node management screen
- Fixed scaling of icons in inventory editor
- Moved hotspot toolbar to right of hotspot image in large screen widths
- Changed player inventory update indicator
- Updated require item button
- Added message in required item tooltip as to whether an answer is visible or not
- Reduce inventory preview in creator to simple Edit Inventory Items (# of items)

To do:
- Hotspot polygon creator broke

clpetersonucf commented on src/creator.html:
```diff
@@ -46,7 +47,6 @@ <h1 id="title" ng-click="showTitleEditor = true">{{title}}</h1>
 				<p>Widget Scoring Mode</p>
 				<button type="button" class="header-options-button" ng-click="showScoreModeDialog = true; showBackgroundCover = true">{{scoreMode}}</button>
 			</div>
-			<!-- <button type="button" class="temp-qset-gen-button" ng-click="onSaveClicked('publish')">Simulate Publish</button> -->
```
Generally I like removing unnecessary commented code, but I actually like keeping this one here. It's very useful for debugging.

clpetersonucf commented on src/creator.html:
```diff
@@ -76,6 +76,120 @@ <h4>Give Your Widget a Title</h4>
+						<p>{{inventoryItems.length ? 'Edit Inventory Items ('+inventoryItems.length+')': 'Add Inventory Items'}}</p>
+					</div>
+				</button>
+				<!-- <div class="items-drag-box">
```
If this element is no longer required, go ahead and remove it.

clpetersonucf commented on src/demo.json:
```diff
@@ -24,6 +24,33 @@
                   }
                 }
               ],
+              "items": [
```
These don't really add anything to the demo, and I'm fine with the demo foregoing making use of the inventory system - for now at least.

clpetersonucf commented on src/src-assets/creator-assets/controllers.coffee:
```diff
@@ -183,29 +193,78 @@ Adventure.controller "AdventureCtrl", ['$scope', '$filter', '$compile', '$rootSc
+
+		$scope.loadIcons().then((result) ->
+			angular.forEach result.icons, (icon, index) ->
+				$scope.icons[index] = {
```
I'm a little worried about how this works with the scope lifecycle in angular, but assuming it works... it's probably okay. Worst case, we wrap some of this in a `$scope.$apply`, so it doesn't interrupt the scope lifecycle and cause weird race conditions.

cayb0rg replied:
To reduce loading times, I decided to remove the XMLHttpRequest entirely and manually insert the array of icons (generated by the load_icons.sh script)

clpetersonucf commented on src/src-assets/creator-assets/services.coffee:
```diff
@@ -404,6 +462,75 @@ Adventure.service "treeSrv", ['$rootScope','$filter','$sanitize','legacyQsetSrv'
 
 		return
 
+	updateAllItems = (tree, updatedItem) ->
```
This one is confusing me a little bit - I assumed a given node would only need a reference to an item ID, and a "master list" of items would contain each item's actual properties: the name, description, and icon. Are those properties also being stored per-node wherever an item is associated with a node? That seems overly complicated - ideally, the item ID remains static (sort of like node IDs), and only the properties of the item in the "master list" would need updating. Feel free to clarify if I'm wrong here.

clpetersonucf reviewed (changes requested):
Now that we're at the point where we can really start to assess what's ready and what still needs polished, I'm going to be reviewing the code changes a little deeper. There will probably several waves of feedback as I find things and we continue to test and iterate.

Bugs I've discovered:

- Drawing polygons over hotspot images isn't working. If I had to guess, the Canvas element that sits on top of the hotspot is misaligned or obscured in z-index by something else.
- Creating an in-between node on a branch that requires an item (has the lock symbol) causes the creator to break

Other feedback:
- I'm not sure about the new directories being added. There's a `style.css` in the `_screen-shots` directory that should not be there. There's both a `src/assets/font` and `src/assets/fonts` directory, with (I assume) the contents of the former moved to the latter.
- There's now a `src/assets/item-assets` directory, with yet another `fonts` directory and a new set of fonts. I would try and consolidate all of these so there's only one set of icomoon font files and whatever other associated files they require.
- As commented below, we can probably leave the demo alone for now.

clpetersonucf commented:
Two new observations:
- Saving a new instance as a draft and then reloading the creator causes the creator to choke when initializing the existing qset.
- The player appears to be having issues with inventory items now - they are not displayed properly in the inventory screen, and are not referenced properly in answers that require the corresponding item.

FrenjaminBanklin reviewed (approved):
There's a lot of potential in the new additions.

I think the working area in the creator can get a bit cramped - some elements can potentially be scaled down, font sizes shrank, etc. to make a little bit more space, but functionally everything new seems to be working well.

Requiring the usage of the 'advanced options' to specify both a minimum and maximum of zero for any given item instead of just allowing the single 0 feels a bit awkward - Potentially treat a single zero as 'the user must have none of this item' and have an optional toggle for 'the user can have any number of this item, even zero' - though that'd be basically the same as the path just being available at all times.

Otherwise, barring some additional things being worked on in separate PRs, this all looks good to me.

clpetersonucf reviewed (approved):
With a feature this large, there's likely still going to be additional items to address, but we've tested this pretty well and it feels like a really solid foundation to the inventory system. Approved.
